code/modules/mob/living/simple_animal/hostile/regalrat.dm

```diff
@@ -82,16 +84,62 @@
 	else if(user != src && istype(user,/mob/living/simple_animal/hostile/regalrat))
 		. += "<span class='warning'>Who is this foolish false king? This will not stand!</span>"
 
+/mob/living/simple_animal/hostile/regalrat/handle_environment(datum/gas_mixture/environment)
+	. = ..()
+	if(stat == DEAD || !environment || !environment.gases[/datum/gas/miasma])
+		return
+	var/miasma_percentage = environment.gases[/datum/gas/miasma][MOLES] / environment.total_moles()
+	if(miasma_percentage>=0.25)
+		heal_bodypart_damage(1)
+
 /mob/living/simple_animal/hostile/regalrat/AttackingTarget()
+	if (busy)
+		return
 	. = ..()
-	if(istype(target, /obj/item/food/cheesewedge))
+	if(istype(target, /obj/machinery/disposal))
```

Comment:
This entire `istype` if tree is getting pretty large. I think with this many items, I'd prefer this be turned into a signal that the objects register *themselves*, rather than that logic be contained in the regal rat.

Reply:
Similar to heretics rust_act?

I tried doing a switch case. Should that work?

Reply:
That's essentially the same thing as this, it's messy.

What the `AttackingTarget` code should be is simply firing a signal. Everything in this list, such as /obj/machinery/disposal, in their *own code*, will register this signal. That way, instead of code for say, rummaging through disposal bins being in the code for regal rats, it's in the code for disposal bins. Look around in the codebase for `RegisterSignal` and similar in order to get a good idea of what I'm asking.

What I'm asking for isn't totally trivial, feel free to ask in the coding channel on /tg/cord if you need help.

Reply:
I gave it a shot. Take a look!

Something like that?

Reply:
Looks right to me! Had some quick quips though.